nltk/model/counter.py

```diff
@@ -0,0 +1,99 @@
+# Natural Language Toolkit: Language Model Counters
+#
+# Copyright (C) 2001-2016 NLTK Project
+# Author: Ilia Kurenkov <ilia.kurenkov@gmail.com>
+# URL: <http://nltk.org/>
+# For license information, see LICENSE.TXT
+
+from __future__ import unicode_literals
+
+from collections import Counter, defaultdict
+from itertools import chain
+
+from nltk.util import ngrams
+from nltk.probability import FreqDist, ConditionalFreqDist
+from nltk import compat
+
+
+@compat.python_2_unicode_compatible
+class LanguageModelVocabulary(Counter):
+    """Stores language model vocabulary.
+
+    Satisfies two common language modeling requirements for a vocabulary:
+    - When checking membership and calculating its size, filters items by comparing
+      their counts to a cutoff value.
+    - Adds 1 to its size so as to account for "unknown" tokens.
+    """
+
+    def __init__(self, unknown_cutoff, *counter_args):
+        Counter.__init__(self, *counter_args)
+        self.cutoff = unknown_cutoff
+
+    def __contains__(self, item):
+        """Only consider items with counts GE to cutoff as being in the vocabulary."""
+        return self[item] >= self.cutoff
+
+    def __len__(self):
+        """This should reflect a) filtering items by count, b) accounting for unknowns.
+
+        The first is achieved by relying on the membership check implementation.
+        The second is achieved by adding 1 to vocabulary size.
+        """
+        # the if-clause here looks a bit dumb, should we make it clearer?
+        return sum(1 for item in self if item in self) + 1
+
+
+@compat.python_2_unicode_compatible
+class NgramCounter(object):
+    """Class for counting ngrams"""
+
+    def __init__(self, highest_order, unknown_cutoff, training_text,
+                 start_symbol="<s>", end_symbol="</s>", unknown_label="<UNK>"):
+        """
```

Comment:
It might be more intuitive if the `training_text` should be the first param.

```
def __init__(self, training_text, highest_order, unknown_cutoff,
                    start_symbol="<s>", end_symbol="</s>", unknown_label="<UNK>")
```